mrmustard/utils/homodyne.py

```diff
@@ -241,3 +240,142 @@ def _probs_homodyne_mixed(state_dm, hbar):
     probs = out_factor * math.real(math.sum(sum_terms, axes=[1, 2]))
 
     return x, probs
+
+
+def sample_heterodyne_fock(state: State, hbar: float) -> Tuple[float, float, State]:
+    r"""Given a single-mode state, it samples the Husimi function :math:`\pi^{-1} \tr [ \rho |\alpha><\alpha| ]`,
+    where `\rho` is the reduced density matrix of the state on the measured mode and `|\alpha>` is a coherent state.
+
+    Args:
+        state (array): ket or dm of the state being measured
+        hbar: value of hbar
+
+    Returns:
+        tuple(float, float): outcome and probability of the outcome
+    """
+    if len(state.shape) == 1:
+        Q, P, probs = _probs_heterodyne_pure(state, hbar)
+    elif len(state.shape) == 2:
+        Q, P, probs = _probs_heterodyne_mixed(state, hbar)
+    else:
+        raise ValueError(
+            "Input state has dimension {state.shape}. Make sure is either a single mode ket or dm."
+        )
+
+    # draw a sample from the distribution
+    probs = probs.flatten()
+    Q = Q.flatten()
+    P = P.flatten()
+    pdf = math.Categorical(probs=probs, name="heterodyne_dist")
+    sample_idx = pdf.sample()
+    Q_sample = math.gather(Q, sample_idx)
+    P_sample = math.gather(P, sample_idx)
+    probability_sample = math.gather(probs, sample_idx)
+
+    return Q_sample, P_sample, probability_sample
+
+
+def _probs_heterodyne_pure(state_ket, hbar):
+
+    cutoff = state_ket.shape[0]
+
+    # calculate prefactors of the PDF
+    q_tensor = math.new_constant(estimate_quadrature_axis(cutoff), "q_tensor")
+    x = np.sqrt(hbar) * q_tensor
+    p = hbar * q_tensor
+
+    Q, P, probs = husimi_pure(state_ket.numpy(), x.numpy(), p.numpy(), hbar)
+
+    return Q, P, probs
+
+
+def _probs_heterodyne_mixed(state_dm, hbar):
+
+    cutoff = state_dm.shape[0]
+
+    # calculate prefactors of the PDF
+    q_tensor = math.new_constant(estimate_quadrature_axis(cutoff), "q_tensor")
+    x = np.sqrt(hbar) * q_tensor
+    p = hbar * q_tensor
+
+    Q, P, probs = husimi_mixed(state_dm.numpy(), x.numpy(), p.numpy(), hbar)
+
+    return Q, P, probs
+
+
+@njit
```

Comment:
Maybe we can put husimi in the same place as the new Wigner module? Like a quasiprobability module?

Reply:
done!